### PR TITLE
fix: gogov error handling

### PR DESCRIPTION
**Pull request by foochifa**
## Problem

Closes FRM-901

## Solution

Broke down the GoGov errors based on the discussion in the linear link and from #6455 

Additionally have also identified an edge case of `BAD_REQUEST` that should be handled as a 500 error. Wherein if there was a HTTPS link error, which means that our `process.env.APP_URL` is configured incorrectly.

Lastly, have checked with GoGov Team on error shapes, seems that there was an intention to expand the error type but it wasn't prioritised, will try and make the changes for them instead and then update our error handling in `admin-form.utils.ts`

**Breaking Changes** 

- [ ] No - this PR is backwards compatible  

## Before & After Screenshots

**BEFORE**:

**AFTER**:

Success
image

Short link taken
image

Validation error
image

Any other unexpected issues (429, 5xx)
image

## Tests

- [ ] Try to create a GoGov link with an existing short link
- [ ] It should fail and return 400
- [ ] FE modal should show short link is taken
- [ ] Try to create a GoGov link with invalid characters like underscore or capital letters
- [ ] It should fail and return 400
- [ ] FE modal should show validation error
- [ ] Go to Param store and delete GoGov API key
- [ ] Try to create a GoGov link with a unique string
- [ ] It should fail and return 500
- [ ] FE modal should show an unexpected error
- [ ] Restore the API key

Regression
- [ ] Try to create a GoGov link with a unique string
- [ ] The link should be successfully created

### CODE REVIEW

**Comment by foochifa** [comment]
@LinHuiqing will tag you first! ~~But will add in the manual test later~~ Manual tests added!

@staceytan1998 do help me check and update the copies of the error messages! 

Thank you!!

File: frontend/src/features/admin-form/share/ShareFormModal.tsx [hunk 1]
```diff
@@ -201,7 +208,17 @@ export const ShareFormModal = ({
...
-      setGoLinkHelperText(goLinkClaimFailureHelperText)
+
+      let errMessage =
+        'Something went wrong. Try refreshing this page. If this issue persists, contact support@form.gov.sg.'
+
+      if (err instanceof HttpError && err.code === StatusCodes.BAD_REQUEST)
+        errMessage =
+          err.message === GO_VALIDATION_ERROR_MESSAGE
+            ? 'Short links should only consist of lowercase letters, numbers and hyphens.'
+            : 'Short link is already in use.'
```
Comment (LinHuiqing) [inline frontend/src/features/admin-form/share/ShareFormModal.tsx]: Could you refactor this as a case switch for it to be cleaner? So err.message should be `GO_VALIDATION_ERROR_MESSAGE | GO_ALREADY_EXIST_ERROR_MESSAGE` typed

File: frontend/src/features/admin-form/share/ShareFormModal.tsx [hunk 2]
```diff
@@ -201,7 +208,17 @@ export const ShareFormModal = ({
       return
     } catch (err) {
       setClaimGoLoading(false)
-      setGoLinkHelperText(goLinkClaimFailureHelperText)
+
+      let errMessage =
```
Comment (LinHuiqing) [inline frontend/src/features/admin-form/share/ShareFormModal.tsx]: Hmm is this error message not a constant somewhere in our code base yet? 👀 
Oh, its not, interesting HAHA

File: src/app/modules/form/admin-form/admin-form.utils.ts [hunk 3]
```diff
@@ -517,3 +547,41 @@ export const verifyUserBetaflag = (
...
+
+  const responseData = error.response?.data as GoGovReturnedData
+
+  switch (error.response?.status) {
+    case StatusCodes.BAD_REQUEST:
+      // There can be three types of Bad Request from GoGov
+      // Short link already exists, which returns type=ShortUrlError
+      // Or validation error, which does not contain type
+      // Or if url is not https (like localhost), however, this should not happen as we prepend the app url in admin-form-controller
+      // TODO: Verify with GoGov team on response data shape
```
Comment (LinHuiqing) [inline src/app/modules/form/admin-form/admin-form.utils.ts]: This has been done right? Based on the PR description

File: src/app/modules/form/admin-form/admin-form.utils.ts [hunk 4]
```diff
@@ -517,3 +547,41 @@ export const verifyUserBetaflag = (
...
+      return responseData.type
+        ? new GoGovAlreadyExistError()
+        : responseData.message.includes(urlFormatError)
+        ? new GoGovServerError(
+            'GoGov server returned 400 for URL formatting error',
+          )
+        : new GoGovValidationError()
```
Comment (LinHuiqing) [inline src/app/modules/form/admin-form/admin-form.utils.ts]: nit: can this be in the same order as what you commented? either by changing this or the comments

**Review by LinHuiqing — COMMENTED** [review]
Thanks @foochifa for working on this! Will test it now

File: frontend/src/features/admin-form/share/ShareFormModal.tsx [hunk 5]
```diff
@@ -370,11 +387,12 @@ export const ShareFormModal = ({
                         )}
                       </Stack>
                       {goLinkHelperText && (
+                        // padding on icon box to emulate padding from <Text>
                         <FormHelperText color={goLinkHelperText.color}>
-                          <Stack direction="row" align="center">
-                            <Box>{goLinkHelperText.icon}</Box>
+                          <HStack alignItems="flex-start">
```
Comment (LinHuiqing) [inline frontend/src/features/admin-form/share/ShareFormModal.tsx:407]: Ooh I'm curious, is there a difference between this and Stack? Why the change?

**Review by LinHuiqing — APPROVED** [review]
tested! can merge after the comments have been addressed! thanks @foochifa! 🙏

File: src/app/modules/form/admin-form/admin-form.utils.ts (hunk 3 above)
Comment (foochifa) [reply src/app/modules/form/admin-form/admin-form.utils.ts]: Yep! Now the todo is to modify this when GoGov addressed their legacy issue on data shape

File: frontend/src/features/admin-form/share/ShareFormModal.tsx (hunk 2 above)
Comment (foochifa) [reply frontend/src/features/admin-form/share/ShareFormModal.tsx]: 😅 perhaps let me make all 3 errors a constant in the ShareFormModal itself

File: frontend/src/features/admin-form/share/ShareFormModal.tsx (hunk 5 above)
Comment (foochifa) [reply frontend/src/features/admin-form/share/ShareFormModal.tsx:407]: Oh its just a horizontal stack 😅 HAHAHA I was just playing around with different types of flex and stacks. HStack is literally <Stack direction='row>. I can change it back hahaha

Comment (LinHuiqing) [reply frontend/src/features/admin-form/share/ShareFormModal.tsx:407]: Oh no, no need, was just wondering if there was a best practice or optimisation thing or something! Can leave it haha no worries! This comment is just for my learning 😂

File: frontend/src/features/admin-form/share/ShareFormModal.tsx (hunk 2 above)
Comment (LinHuiqing) [reply frontend/src/features/admin-form/share/ShareFormModal.tsx]: Oh this is a very minor nit btw, I don't feel strongly abt it so don't worry abt it! Was just wondering lel

File: src/app/modules/form/admin-form/admin-form.utils.ts (hunk 3 above)
Comment (foochifa) [reply src/app/modules/form/admin-form/admin-form.utils.ts]: [a6c20ea](https://github.com/opengovsg/FormSG/pull/6714/commits/a6c20ea7794bf2452514eb34e2d2e822a81d84d4)

File: src/app/modules/form/admin-form/admin-form.utils.ts (hunk 4 above)
Comment (foochifa) [reply src/app/modules/form/admin-form/admin-form.utils.ts]: [a6c20ea](https://github.com/opengovsg/FormSG/pull/6714/commits/a6c20ea7794bf2452514eb34e2d2e822a81d84d4)

File: frontend/src/features/admin-form/share/ShareFormModal.tsx (hunk 2 above)
Comment (foochifa) [reply frontend/src/features/admin-form/share/ShareFormModal.tsx]: [b9ebcd9](https://github.com/opengovsg/FormSG/pull/6714/commits/b9ebcd948a2395f29eb5cac71b4f32f4037c7535)

I think still good to not have magic floating texts haha so abstracted them

File: frontend/src/features/admin-form/share/ShareFormModal.tsx (hunk 1 above)
Comment (foochifa) [reply frontend/src/features/admin-form/share/ShareFormModal.tsx]: [b9ebcd9](https://github.com/opengovsg/FormSG/pull/6714/commits/b9ebcd948a2395f29eb5cac71b4f32f4037c7535)
